### ISSUE TITLE
Expo with RxDB Premium SQLite, any way to make it work

### ISSUE THREAD

**Issue by jwallet** [pubkey/rxdb#4536]
Hello pubkey

since the source code is hidden and you are probably the only one that can answer a premium related question,

how can I make the getter `getSQLiteBasicsExpoSQLite` works with any sqlite database on expo for react-native on android?

I do not want to use the pouchdb-adapter for sqlite and I tried to setup the SQLite Premium plugin with pretty much every sqlite database without success. your react-native example is running with a database in-memory which is only good for testing. your rxdb-premium sqllite says it works on react-native, I do hope it works on expo in any way.

```
 switch (storageKey) {
    case 'expo':
      return getRxStorageSQLite({
        sqliteBasics: getSQLiteBasicsExpoSQLite(exposqlite.openDatabase),
        log: console.log.bind(console),
      });
    case 'sqlite2':
      return getRxStorageSQLite({
        sqliteBasics: getSQLiteBasicsWebSQL(sqlite2.openDatabase),
        log: console.log.bind(console),
      });
    case 'sqlitestorage': {
      sqlitestorage.enablePromise(true);
      return getRxStorageSQLite({
        sqliteBasics: getSQLiteBasicsWebSQL(sqlitestorage.openDatabase),
        log: console.log.bind(console),
      });
    }
    case 'quicksqlite':
      return getRxStorageSQLite({
        sqliteBasics: getSQLiteBasicsQuickSQLite(quicksqlite.open),
        log: console.log.bind(console),
      });
    default:
      throw new Error(`RxDB: Crashing... Storage key "${storageKey}" is not supported.`);
  }
```

you can see the stackoverflow post [here](https://stackoverflow.com/questions/75677335/any-way-to-make-rxdb-work-for-react-native-on-expo-with-a-sqlite-database-i-own)

**Comment by pubkey** [comment]
From the [docs](https://rxdb.info/rx-storage-sqlite.html)
> [expo-sqlite](https://www.npmjs.com/package/expo-sqlite) cannot be used on android (but it works on iOS) because it uses an [outdated SQLite version](https://expo.canny.io/feature-requests/p/expo-sqlite-ship-newer-sqlite3-version-on-android)

So you can use expo-sqlite in iOS, but not on android. Last week there where some problems fixed with `getSQLiteBasicsWebSQL()`  that can work for you with [react-native-sqlite-2](https://www.npmjs.com/package/react-native-sqlite-2)

**Comment by jwallet** [comment]
Thanks pubkey,

Yes, I've seen the documentation page. But, we did hope during the POC, that there was a way to make it work on both android and iOS without ditching expo. Looks like this guy in the issue you linked had more success than me. I still end up with an error with https://github.com/craftzdog/react-native-sqlite-2/issues/22 like mentioned in the stack overflow post

```
{"name":"TypeError","message":"null is not an object (evaluating 'RNSqlite2.exec')"
```

some say in that issue that is just incompatible with expo.

it's kind of a let down if the Premium SQLite plugin doesn't even work on android and expo, considering how big expo is in the community. If there is no proper way to set it up with the latest sdk 48, I guess we will try to make it work on rxdb-classic using expo-sqlite with memory storage, will see.

**Comment by pubkey** [comment]
Can you enable the logging, this might gave a clue on what is wrong.

```ts
const storage = getRxStorageSQLite({
  sqliteBasics: getSQLiteBasicsWebSQL(SQLite.openDatabase),
  // pass log function
  log: console.log.bind(console)
});
```

**Comment by jwallet** [comment]
Ok, so that's what I get when I log it:

```bash
 LOG  ## RxStorage SQLite log: open(0) guay_sqlite_42a37798-6e4c-44b0-a0a1-2590a0431755_1678135011_1678460000420
 LOG  ## RxStorage SQLite log: open(0) DONE
 LOG  ## RxStorage SQLite log: run(1) {"query":"BEGIN;","params":[]}
# some logs outside of rxdb
 LOG  ## RxStorage SQLite log: run(1593) {"query":"BEGIN;","params":[]}
 LOG  ## RxStorage SQLite log: run(1593) ERROR
 LOG  open transaction error (will retry):
 LOG  {"name":"TypeError","message":"null is not an object (evaluating 'RNSqlite2.exec')","stack":"exec@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:265791:18\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:264884:13\ntryCallTwo@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26518:9\ndoResolve@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26650:25\nPromise@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26537:14\ny@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:264883:23\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:264849:18\ngeneratorResume@[native code]\nasyncGeneratorStep@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21753:26\n_next@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21772:29\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21777:14\ntryCallTwo@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26518:9\ndoResolve@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26650:25\nPromise@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26537:14\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21769:25\nrun@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:264852:29\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:263988:59\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:264147:22\ngeneratorResume@[native code]\nasyncGeneratorStep@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21753:26\n_next@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:21772:29\ntryCallOne@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26510:16\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:26591:27\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:27535:26\n_callTimer@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:27450:17\n_callReactNativeMicrotasksPass@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:27483:17\ncallReactNativeMicrotasks@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:27647:44\n__callReactNativeMicrotasks@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:20872:46\nhttp://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:20672:45\n__guard@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:20852:15\nflushedQueue@http://192.168.2.12:19000/index.bundle?platform=android&dev=true&hot=false:20671:21\nflushedQueue@[native code]\ncallFunctionReturnFlushedQueue@[native code]"}
# it then repeats indefinitely the last 4 logs
```

**Comment by kilbot** [comment]
Hi @jwallet, I am also looking for a good DB option to work with expo. I was actually thinking about creating my own [Native Module](https://docs.expo.dev/modules/native-module-tutorial/), I think Expo have made some improvements to the API recently which makes easier 😬

If we created our own module we could [choose a SQLite library](https://reactnative.directory/?search=sqlite) that best works with RxDB, or perhaps another DB? Would you be interested in development something together?

**Comment by jwallet** [comment]
@kilbot you can probably look at this repo, that guy tried to fork expo-sqlite and update the android database to use requery.sqlite-android https://github.com/Joeman29/expo-sqlite-update

**Comment by pubkey** [comment]
@jwallet  from your logs it looks like you are able to create the database connection but running the queries fails.
Can you try using plain SQLite without rxdb and check if there the querying works for you?

**Comment by jwallet** [comment]
@pubkey  I tried it in a new env and my expo-sqlite (updated the android database to support JSON_EXTRACT) is working using my npm package fine, but when I tried it in my project it failed on me, a different log than the one I got from sqlite2 but really similar . So I guess both should work just fine. 

I will check my expo configs. The stack is just so big, but I do know it fails even if in App I just open the simplest sqlite database setup in a hook. So probably not caused by the React code itself. Anyway, close it or keep it open, I will come back later here to leave what is causing the issue , if it can help someone else.

**Comment by jwallet** [comment]
So I found that `expo start` install new libs like `expo-splash-screen`, it's convenient because you don't have to wait for the android app to build for 5 minutes, it just uses the expo libs loaded in expo-go. But, I need to build it, so I looked at the command `npx expo run:android` and it updated some core expo libs, and removed others, like it was fixing my `package.json` dependencies. It did build the android, but before I was not able to load `@expo/websql` in my `expo-sqlite` fork, it probably fixed that dependencie, cause now I have result.

I did come with a new issue while it was syncing.
